src/dbus-objects/addrconf.c

```diff
@@ -636,80 +614,348 @@ ni_objectmodel_addrconf_forward_request(ni_dbus_addrconf_forwarder_t *forwarder,
+	return act;
+}
+
+static ni_addrconf_drop_collect_t *
+ni_addrconf_drop_collect_cast(ni_addrconf_action_t *action)
+{
+	if (!action || action->exec != ni_addrconf_drop_collect_exec)
+		return NULL;
+	return (ni_addrconf_drop_collect_t *)action;
+}
+
+static inline ni_bool_t
+ni_addrconf_drop_collect_test(ni_addrconf_updater_t *updater, ni_addrconf_drop_collect_t *collect)
+{
+	if (collect->complete)
+		return TRUE;
+
+	if (collect->deadline && !ni_lifetime_left(collect->deadline, &updater->astart, NULL))
+		return TRUE;
+
+	return FALSE;
+}
+
+static int
+ni_addrconf_drop_collect_exec(ni_netdev_t *dev, ni_addrconf_lease_t *lease)
+{
+	ni_addrconf_drop_collect_t *collect;
+
+	if (!dev || !lease || !lease->updater)
+		return -1;
+
+	if (!(collect = ni_addrconf_drop_collect_cast(lease->updater->action)))
+		return -1;
+
+	if (ni_addrconf_drop_collect_test(lease->updater, collect)) {
+		lease->state = NI_ADDRCONF_STATE_RELEASED;
+		return 0;
+	}
+	return 1;
+}
+
+static inline void
+ni_addrconf_drop_collect_free(ni_addrconf_drop_collect_t *act)
+{
+	/* nothing to free inside */
+	free(act);
+}
+
+static ni_addrconf_drop_collect_t *
+ni_addrconf_drop_collect_new(void)
+{
+	static const char *info = "collecting drop event";
+	ni_addrconf_drop_collect_t *act;
+
+	if (!(act = calloc(1, sizeof(*act))))
+		return NULL;
+
+	act->deadline = 10; /* worst case 10 sec no-reply deadline
+			       (buggy/stopped/crashed supplicant) */
+	act->action.info = info;
+	act->action.exec = ni_addrconf_drop_collect_exec;
+	return act;
+}
+
+static ni_bool_t
+ni_addrconf_drop_actions_arm(ni_dbus_addrconf_forwarder_t *forwarder,
+				ni_addrconf_updater_t *updater,
+				const ni_dbus_variant_t *dict)
+{
+	ni_addrconf_drop_forward_t *forward = NULL;
+	ni_addrconf_drop_collect_t *collect = NULL;
+	ni_addrconf_action_t *pos;
+
+	if (!updater)
+		return FALSE;
+
+	if (!(forward = ni_addrconf_drop_forward_new(forwarder)))
+		goto failure;
+
+	ni_addrconf_drop_args_from_variant(forwarder, &forward->args, dict);
+
+	if (!(collect = ni_addrconf_drop_collect_new()))
+		goto failure;
+
+	/*
+	 * Insert the forward action behind address removal
+	 * and append the collect action as the last action.
+	 */
+	pos = ni_addrconf_action_list_find_exec(updater->action,
+				ni_addrconf_action_addrs_remove);
+
+	if (pos && ni_addrconf_action_list_insert(&pos->next, &forward->action)) {
+		if (ni_addrconf_action_list_append(&pos->next, &collect->action))
+			return TRUE;
+	}
+
+failure:
+	ni_addrconf_drop_collect_free(collect);
+	ni_addrconf_drop_forward_free(forward);
+	return FALSE;
+}
+
+static ni_addrconf_action_t *
+ni_addrconf_drop_collect_complete(ni_addrconf_lease_t *lease)
+{
+	ni_addrconf_drop_collect_t *collect;
+	ni_addrconf_action_t *action;
+
+	if (!lease || !lease->updater)
+		return NULL;
+
+	action = ni_addrconf_action_list_find_exec(lease->updater->action,
+					ni_addrconf_drop_collect_exec);
+	if (!(collect = (ni_addrconf_drop_collect_t *)action))
+		return NULL;
```

Comment:
Just a question:
This code expect only one of the `ni_addrconf_drop_collect_t*` objects inside the action list. But I do not see a guard when we adding this type to the list. Do we need to guard it?

Reply:
The guard is the `ni_addrconf_drop_collect_exec` -- we search for the collect action using it. When we don't find it, `action == NULL` (and also the cast to collect). The above code could be also:
```C
action = ni_addrconf_action_list_find_exec(lease->updater->action, ni_addrconf_drop_collect_exec);
if (!action)
	return NULL;
collect = (ni_addrconf_drop_collect_t *)action;
```
It's just written in one line.

Reply:
> expect only one of the ni_addrconf_drop_collect_t* objects inside the action list

Ahh.. now I understand, you mean there are multiple collect actions in  the action list like:
   ```updater->actions->[collect]->..->[collect]->..```
and we complete only the 1st one, so the  2nd would continue to collect/wait.

No, this can't happen. We always allocate new releasing lease and a new updater with a fresh action list where we add  only 1 forward action behind `remove address` (see arm funct) and 1 collect action as last action in the list.